File: frontend/nextjs-app/scripts/gen-hero-subtitle-paths.mjs
/**
 * Caveat TTF로 부제 문장의 글자별 SVG path 생성 → src/data/textSubtitlePaths.json
 * 사용: `npm i -D opentype.js` 후 `node scripts/gen-hero-subtitle-paths.mjs`
 * (CI/Docker는 package-lock 동기화를 위해 npm 스크립트는 Python 생성기만 사용)
 */
import opentype from "opentype.js";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const FONT_URL =
  "https://raw.githubusercontent.com/google/fonts/main/ofl/caveat/Caveat%5Bwght%5D.ttf";
const TEXT = "A personal log across space and time.";
const OUT = path.join(__dirname, "../src/data/textSubtitlePaths.json");

const fontSize = 52;
const baselineY = fontSize * 0.72;

const res = await fetch(FONT_URL);
if (!res.ok) throw new Error(`Font fetch failed: ${res.status}`);
const buf = Buffer.from(await res.arrayBuffer());
const font = opentype.parse(buf);
const scale = fontSize / font.unitsPerEm;

let x = 0;
let prevGlyph = null;
const byChar = [];
let minX = Infinity;
let minY = Infinity;
let maxX = -Infinity;
let maxY = -Infinity;

function expandBox(bb) {
  if (!bb || bb.x1 === undefined || !Number.isFinite(bb.x1)) return;
  minX = Math.min(minX, bb.x1);
  minY = Math.min(minY, bb.y1);
  maxX = Math.max(maxX, bb.x2);
  maxY = Math.max(maxY, bb.y2);
}

for (let i = 0; i < TEXT.length; i++) {
  const glyph = font.charToGlyph(TEXT[i]);
  if (prevGlyph && glyph) {
    x += font.getKerningValue(prevGlyph, glyph) * scale;
  }
  const gPath = glyph.getPath(x, baselineY, fontSize);
  const d = gPath.toPathData(2);
  if (d && d.trim().length > 2) {
    byChar.push({ d });
    expandBox(gPath.getBoundingBox());
  } else {
    byChar.push(null);
  }
  x += glyph.advanceWidth * scale;
  prevGlyph = glyph;
}

const pad = fontSize * 0.2;
if (!Number.isFinite(minX)) {
  minX = 0;
  minY = 0;
  maxX = x;
  maxY = fontSize;
}
const viewBox = `${(minX - pad).toFixed(2)} ${(minY - pad).toFixed(2)} ${(maxX - minX + 2 * pad).toFixed(2)} ${(maxY - minY + 2 * pad).toFixed(2)}`;

fs.writeFileSync(OUT, JSON.stringify({ viewBox, byChar }, null, 2), "utf8");
console.log("Wrote", OUT, "chars:", byChar.length, "drawable:", byChar.filter(Boolean).length);

File: frontend/nextjs-app/scripts/gen-hero-subtitle-paths.py
#!/usr/bin/env python3
"""Caveat(가변 TTF)로 부제 문장의 글자별 SVG path → src/data/textSubtitlePaths.json

  pip install fonttools
  curl -sL -o /tmp/Caveat.ttf \
    'https://raw.githubusercontent.com/google/fonts/main/ofl/caveat/Caveat%5Bwght%5D.ttf'
  python3 scripts/gen-hero-subtitle-paths.py
"""
from __future__ import annotations

import json
import math
import os
import sys
from pathlib import Path

from fontTools.misc.transform import Transform
from fontTools.pens.boundsPen import BoundsPen
from fontTools.pens.svgPathPen import SVGPathPen
from fontTools.pens.transformPen import TransformPen
from fontTools.ttLib import TTFont

TEXT = "A personal log across space and time."
FONT_SIZE = 52
BASELINE_Y = FONT_SIZE * 0.72
PAD = FONT_SIZE * 0.2

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "src" / "data" / "textSubtitlePaths.json"
FONT_CANDIDATES = [
    Path("/tmp/Caveat.ttf"),
    ROOT / "scripts" / "Caveat.ttf",
]


def main() -> None:
    font_path = next((p for p in FONT_CANDIDATES if p.is_file()), None)
    if not font_path:
        print("폰트 없음. 다음 중 하나를 두세요:", file=sys.stderr)
        for p in FONT_CANDIDATES:
            print(f"  {p}", file=sys.stderr)
        sys.exit(1)

    font = TTFont(str(font_path))
    glyph_set = font.getGlyphSet()
    cmap = font["cmap"].getBestCmap()
    hmtx = font["hmtx"].metrics
    upem = font["head"].unitsPerEm
    scale = FONT_SIZE / upem

    x = 0.0
    by_char: list[dict | None] = []
    min_x = min_y = math.inf
    max_x = max_y = -math.inf

    for ch in TEXT:
        gname = cmap.get(ord(ch))
        if not gname:
            by_char.append(None)
            continue

        glyph = glyph_set[gname]
        t = Transform(scale, 0, 0, -scale, x, BASELINE_Y)

        bp = BoundsPen(glyph_set)
        glyph.draw(TransformPen(bp, t))
        if bp.bounds is not None:
            x0, y0, x1, y1 = bp.bounds
            min_x = min(min_x, x0)
            min_y = min(min_y, y0)
            max_x = max(max_x, x1)
            max_y = max(max_y, y1)

        pen = SVGPathPen(glyph_set)
        glyph.draw(TransformPen(pen, t))
        d = pen.getCommands().strip()
        if d:
            by_char.append({"d": d})
        else:
            by_char.append(None)

        adv, _ = hmtx[gname]
        x += adv * scale

    if not math.isfinite(min_x):
        min_x, min_y, max_x, max_y = 0, 0, x, FONT_SIZE

    view_box = f"{min_x - PAD:.2f} {min_y - PAD:.2f} {max_x - min_x + 2 * PAD:.2f} {max_y - min_y + 2 * PAD:.2f}"
    payload = {"viewBox": view_box, "byChar": by_char}
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    drawn = sum(1 for c in by_char if c)
    print(f"Wrote {OUT} (chars={len(by_char)}, paths={drawn})")


if __name__ == "__main__":
    main()
